crates/dslab-iaas/tests/test_core.rs

```diff
@@ -0,0 +1,309 @@
+        cloud_sim.spawn_vm_now(
+            i,
+            10,
+            10,
+            100.0,
+            Box::new(ConstLoadModel::new(1.0)),
+            Box::new(ConstLoadModel::new(1.0)),
+            s,
+        );
+        cloud_sim.step_for_duration(5.);
+    }
+
+    let current_time = cloud_sim.current_time();
+    assert_eq!(current_time, 55.);
+    assert_eq!(cloud_sim.host(h1).borrow_mut().get_cpu_load(current_time), 1.0);
+    assert_eq!(cloud_sim.host(h2).borrow_mut().get_cpu_load(current_time), 0.1);
+}
+
+#[test]
+// Can pack 94 VMS despite their total SLA is 94 times bigger than host capacity
+fn test_overcommit() {
+    let sim = Simulation::new(123);
+    let sim_config = SimulationConfig::from_file(&name_wrapper("config_with_overcommit.yaml"));
+    let mut cloud_sim = CloudSimulation::new(sim, sim_config.clone());
+
+    let h = cloud_sim.add_host("h", 200, 200);
+    let s = cloud_sim.add_scheduler("s", Box::new(BestFitThreshold::new(1.0)));
+
+    for i in 1..95 {
+        cloud_sim.spawn_vm_now(
+            i,
+            100,
+            100,
+            1000.0,
+            Box::new(ConstLoadModel::new(0.01)),
+            Box::new(ConstLoadModel::new(0.01)),
+            s,
+        );
+        cloud_sim.step_for_duration(1.);
+    }
+
+    cloud_sim.step_for_duration(5.);
+    let current_time = cloud_sim.current_time();
+    assert_eq!(current_time, 99.);
+    assert_eq!(cloud_sim.host(h).borrow_mut().get_cpu_load(current_time), 0.47);
+}
+
+pub struct BadScheduler {
+    choice: u32,
+}
+
+impl BadScheduler {
+    pub fn new(choice: u32) -> Self {
+        Self { choice }
+    }
+}
+
+impl VMPlacementAlgorithm for BadScheduler {
+    fn select_host(
+        &self,
+        _alloc: &Allocation,
+        _pool_state: &ResourcePoolState,
+        _monitoring: &Monitoring,
+    ) -> Option<u32> {
+        return Some(self.choice);
+    }
+}
+
+#[test]
+// User packs the VM on overloaded host, but the issue is resolved by placement store
+fn test_wrong_decision() {
+    let sim = Simulation::new(123);
+    let sim_config = SimulationConfig::from_file(&name_wrapper("config.yaml"));
+    let mut cloud_sim = CloudSimulation::new(sim, sim_config.clone());
+
+    let h1 = cloud_sim.add_host("h1", 100, 100);
+    let h2 = cloud_sim.add_host("h2", 100, 100);
+    let s = cloud_sim.add_scheduler("s", Box::new(FirstFit::new()));
+
+    cloud_sim.spawn_vm_now(
+        1,
+        100,
+        100,
+        100.0,
+        Box::new(ConstLoadModel::new(1.)),
+        Box::new(ConstLoadModel::new(1.)),
+        s,
+    );
+    cloud_sim.step_for_duration(5.);
+
+    // now host one is overloaded
+    let mut current_time = cloud_sim.current_time();
+    assert_eq!(current_time, 5.);
+    assert_eq!(cloud_sim.host(h1).borrow_mut().get_cpu_load(current_time), 1.);
+    assert_eq!(cloud_sim.vm_status(1), VmStatus::Running);
+
+    let bad_s = cloud_sim.add_scheduler("bad_s", Box::new(BadScheduler::new(3)));
```

Comment:
```suggestion
    let bad_s = cloud_sim.add_scheduler("bad_s", Box::new(BadScheduler::new(h1)));
```

Здесь и далее используйте вместо чисел id хостов.

Reply:
Не исправлено, проверьте